velox/docs/develop/spilling.rst

```diff
@@ -0,0 +1,480 @@
+state for output. The intermediate state of a group by key can be spilled
+multiple times during the operator’s execution. Also note that the sort is
+based on the group by key columns.
+
+OrderBy
+-------
+The order by operator stores all the input rows in a row container and sorts
+them all once after it has received all the inputs. When spilling gets
+triggered, the Spiller will collect a sufficient number of rows to spill to
+satisfy the spill target. Unlike the hash aggregation spilling, we don’t
+partition the rows for spilling as the order by operator needs to generate the
+total order on all the input rows for output. After the spilling completes, the
+operator continues the input processing until the next spill run gets triggered
+which repeats the same process above.
+
+After processing all the inputs, the order by operator first sorts any rows
+left in the row container as a single sorted run, and each spill file on disk
+is also a sorted run. Then the operator creates a single sort merge reader with
+all the sorted runs to produce the final sorted output. Note that the sort here
+needs to use the comparison options specified by the query plan node.
+
+Hash Join
+---------
+
+The hash join in Velox is implemented by hash build and hash probe two types of
+operators, and each belonging to a separate driver pipeline and the two
+pipelines are connected through a shared hash join bridge data structure. The
+hash build operators take the build side (or the right side in a join) inputs
+to build a hash table. After the build completes, one of the hash build
+operators sends the built table to all the hash probe operators through the
+shared hash join bridge. The hash probe operator takes the probe side (or the
+left side of a join) inputs to join with the hash table with one batch at a
+time.
+
+As we can see, the hash probe operator will hold at most one batch of probe
+input rows in memory at a time so the hash probe processing won’t use too much
+memory. The hash build operators might use a large amount of memory to build
+the hash table and keep it in the memory for the entire hash join processing.
+Similar to the order by processing, each hash build operator will first store
+the build side inputs into a row container, and after all the hash build
+operators have processed the inputs, one of them will build a single aggregated
+hash table with the rows collected from all the hash build operators.
+
+.. image:: images/spill-hash-join-build.png
+   :width: 800
+   :align: center
+
+To prevent hash join from running out of memory, the hash build operators need
+to spill part of the build side inputs to disk if the row container grows too
+large during the processing. As mentioned above, the spilling at the hash build
+operators need to coordinate with each other as we want all the operators to
+spill the same set of partitions. Otherwise, if each operator spills
```

Comment:
> Otherwise, if each operator

If operators spill independently, it is possible to end up with all partitions being spilled.